File: app/src/main/java/com/keepassdroid/collections/VariantDictionary.java
/*
 * Copyright 2017 Brian Pellin.
 *
 * This file is part of KeePassDroid.
 *
 *  KeePassDroid is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  KeePassDroid is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePassDroid.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.keepassdroid.collections;

import com.keepassdroid.stream.LEDataInputStream;
import com.keepassdroid.stream.LEDataOutputStream;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

public class VariantDictionary {
    private static final int VdVersion = 0x0100;
    private static final int VdmCritical = 0xFF00;
    private static final int VdmInfo = 0x00FF;

    private Map<String, VdType> dict = new HashMap<String, VdType>();

    private class VdType {
        public static final byte None = 0x00;
        public static final byte UInt32 = 0x04;
        public static final byte UInt64 =0x05;
        public static final byte Bool =0x08;
        public static final byte Int32 =0x0C;
        public static final byte Int64 =0x0D;
        public static final byte String =0x18;
        public static final byte ByteArray =0x42;

        public final byte type;
        public final Object value;

        VdType(byte type, Object value) {
            this.type = type;
            this.value = value;
        }

    }

    private Object getValue(String name) {
        VdType val = dict.get(name);
        if (val == null) {
            return null;
        }

        return val.value;
    }
    private void putType(byte type, String name, Object value) {
        dict.put(name, new VdType(type, value));
    }

    public void setUInt32(String name, long value) { putType(VdType.UInt32, name, value); }
    public long getUInt32(String name) { return (long)dict.get(name).value; }

    public void setUInt64(String name, long value) { putType(VdType.UInt64, name, value); }
    public long getUInt64(String name) { return (long)dict.get(name).value; }

    public void setBool(String name, boolean value) { putType(VdType.Bool, name, value); }
    public boolean getBool(String name) { return (boolean)dict.get(name).value; }

    public void setInt32(String name, int value) { putType(VdType.Int32 ,name, value); }
    public int getInt32(String name) { return (int)dict.get(name).value; }

    public void setInt64(String name, long value) { putType(VdType.Int64 ,name, value); }
    public long getInt64(String name) { return (long)dict.get(name).value; }

    public void setString(String name, String value) { putType(VdType.String ,name, value); }
    public String getString(String name) { return (String)getValue(name); }

    public void setByteArray(String name, byte[] value) { putType(VdType.ByteArray, name, value); }
    public byte[] getByteArray(String name) { return (byte[])getValue(name); }

    public static VariantDictionary deserialize(LEDataInputStream lis) throws IOException {
        VariantDictionary d = new VariantDictionary();

        int version = lis.readUShort();
        if ((version & VdmCritical) > (VdVersion & VdmCritical)) {
            throw new IOException("Invalid format");
        }

        while(true) {
            int type = lis.read();
            if (type < 0) {
                throw new IOException(("Invalid format"));
            }

            byte bType = (byte)type;
            if (bType == VdType.None) {
                break;
            }

            int nameLen = lis.readInt();
            byte[] nameBuf = lis.readBytes(nameLen);
            if (nameLen != nameBuf.length) {
                throw new IOException("Invalid format");
            }
            String name = new String(nameBuf, "UTF-8");

            int valueLen = lis.readInt();
            byte[] valueBuf = lis.readBytes(valueLen);
            if (valueLen != valueBuf.length) {
                throw new IOException("Invalid format");
            }

            switch (bType) {
                case VdType.UInt32:
                    if (valueLen == 4) {
                        d.setUInt32(name, LEDataInputStream.readUInt(valueBuf, 0));
                    }
                    break;
                case VdType.UInt64:
                    if (valueLen == 8) {
                        d.setUInt64(name, LEDataInputStream.readLong(valueBuf, 0));
                    }
                    break;
                case VdType.Bool:
                    if (valueLen == 1) {
                        d.setBool(name, valueBuf[0] != 0);
                    }
                    break;
                case VdType.Int32:
                    if (valueLen == 4) {
                        d.setInt32(name, LEDataInputStream.readInt(valueBuf, 0));
                    }
                    break;
                case VdType.Int64:
                    if (valueLen == 8) {
                        d.setInt64(name, LEDataInputStream.readLong(valueBuf, 0));
                    }
                    break;
                case VdType.String:
                    d.setString(name, new String(valueBuf, "UTF-8"));
                    break;
                case VdType.ByteArray:
                    d.setByteArray(name, valueBuf);
                    break;
                default:
                    assert (false);
                    break;
            }
        }

        return d;
    }

    public static void serialize(VariantDictionary d, LEDataOutputStream los) throws IOException{
        if (los == null) {
            assert(false);
            return;
        }

        los.writeUShort(VdVersion);

        for (Map.Entry<String, VdType> entry: d.dict.entrySet()) {
            String name = entry.getKey();
            byte[] nameBuf = null;
            try {
                nameBuf = name.getBytes("UTF-8");
            } catch (UnsupportedEncodingException e) {
                assert(false);
                throw new IOException("Couldn't encode parameter name.");
            }

            VdType vd = entry.getValue();

            los.write(vd.type);
            los.writeInt(nameBuf.length);
            los.write(nameBuf);

            byte[] buf;
            switch (vd.type) {
                case VdType.UInt32:
                    los.writeInt(4);
                    los.writeUInt((long)vd.value);
                    break;
                case VdType.UInt64:
                    los.writeInt(8);
                    los.writeLong((long)vd.value);
                    break;
                case VdType.Bool:
                    los.writeInt(1);
                    byte bool = (boolean)vd.value ? (byte)1 : (byte)0;
                    los.write(bool);
                    break;
                case VdType.Int32:
                    los.writeInt(4);
                    los.writeInt((int)vd.value);
                    break;
                case VdType.Int64:
                    los.writeInt(8);
                    los.writeLong((long)vd.value);
                    break;
                case VdType.String:
                    String value = (String)vd.value;
                    buf = value.getBytes("UTF-8");
                    los.writeInt(buf.length);
                    los.write(buf);
                    break;
                case VdType.ByteArray:
                    buf = (byte[])vd.value;
                    los.writeInt(buf.length);
                    los.write(buf);
                    break;
                default:
                    assert(false);
                    break;
            }
        }

        los.write(VdType.None);

    }

    public void copyTo(VariantDictionary d) {
       for (Map.Entry<String, VdType> entry : d.dict.entrySet()) {
           String key = entry.getKey();
           VdType value = entry.getValue();

           dict.put(key, value);
       }
    }
;
    public int size() {
        return dict.size();
    }
}

File: app/src/main/java/com/kunzisoft/keepass/database/crypto/VariantDictionary.kt
/*
 * Copyright 2020 Jeremy Jamet / Kunzisoft.
 *
 * This file is part of KeePassDX.
 *
 *  KeePassDX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePassDX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePassDX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.database.crypto

import com.kunzisoft.keepass.utils.*
import java.io.*
import java.nio.charset.Charset
import java.util.*

open class VariantDictionary {

    constructor()

    constructor(d: VariantDictionary) {
        for ((key, value) in d.dict) {
            dict[key] = value
        }
    }

    private val dict: MutableMap<String, VdType> = HashMap()

    private fun getValue(name: String): Any? {
        return dict[name]?.value ?: return null
    }

    private fun putType(type: Byte, name: String, value: Any) {
        dict[name] = VdType(type, value)
    }

    fun setUInt32(name: String, value: UnsignedInt) {
        putType(VdType.UInt32, name, value)
    }

    fun getUInt32(name: String): UnsignedInt? {
        return dict[name]?.value as UnsignedInt?
    }

    fun setUInt64(name: String, value: UnsignedLong) {
        putType(VdType.UInt64, name, value)
    }

    fun getUInt64(name: String): UnsignedLong? {
        return dict[name]?.value as UnsignedLong?
    }

    fun setBool(name: String, value: Boolean) {
        putType(VdType.Bool, name, value)
    }

    fun getBool(name: String): Boolean? {
        return dict[name]?.value as Boolean?
    }

    fun setInt32(name: String, value: Int) {
        putType(VdType.Int32, name, value)
    }

    fun getInt32(name: String): Int? {
        return dict[name]?.value as Int?
    }

    fun setInt64(name: String, value: Long) {
        putType(VdType.Int64, name, value)
    }

    fun getInt64(name: String): Long? {
        return dict[name]?.value as Long?
    }

    fun setString(name: String, value: String) {
        putType(VdType.String, name, value)
    }

    fun getString(name: String): String? {
        return getValue(name) as String?
    }

    fun setByteArray(name: String, value: ByteArray) {
        putType(VdType.ByteArray, name, value)
    }

    fun getByteArray(name: String): ByteArray? {
        return getValue(name) as ByteArray?
    }

    fun size(): Int {
        return dict.size
    }

    companion object {
        private const val VdVersion = 0x0100
        private const val VdmCritical = 0xFF00
        private const val VdmInfo = 0x00FF
        private val UTF8Charset = Charset.forName("UTF-8")

        @Throws(IOException::class)
        fun deserialize(data: ByteArray): VariantDictionary {
            val inputStream = ByteArrayInputStream(data)
            return deserialize(inputStream)
        }

        @Throws(IOException::class)
        fun serialize(variantDictionary: VariantDictionary): ByteArray {
            val byteArrayOutputStream = ByteArrayOutputStream()
            serialize(variantDictionary, byteArrayOutputStream)
            return byteArrayOutputStream.toByteArray()
        }

        @Throws(IOException::class)
        fun deserialize(inputStream: InputStream): VariantDictionary {
            val dictionary = VariantDictionary()
            val version = inputStream.readBytes2ToUShort()
            if (version and VdmCritical > VdVersion and VdmCritical) {
                throw IOException("Invalid format")
            }
            while (true) {
                val type = inputStream.read()
                if (type < 0) {
                    throw IOException("Invalid format")
                }
                val bType = type.toByte()
                if (bType == VdType.None) {
                    break
                }
                val nameLen = inputStream.readBytes4ToUInt().toKotlinInt()
                val nameBuf = inputStream.readBytesLength(nameLen)
                if (nameLen != nameBuf.size) {
                    throw IOException("Invalid format")
                }
                val name = String(nameBuf, UTF8Charset)
                val valueLen = inputStream.readBytes4ToUInt().toKotlinInt()
                val valueBuf = inputStream.readBytesLength(valueLen)
                if (valueLen != valueBuf.size) {
                    throw IOException("Invalid format")
                }
                when (bType) {
                    VdType.UInt32 -> if (valueLen == 4) {
                        dictionary.setUInt32(name, bytes4ToUInt(valueBuf))
                    }
                    VdType.UInt64 -> if (valueLen == 8) {
                        dictionary.setUInt64(name, bytes64ToULong(valueBuf))
                    }
                    VdType.Bool -> if (valueLen == 1) {
                        dictionary.setBool(name, valueBuf[0] != 0.toByte())
                    }
                    VdType.Int32 -> if (valueLen == 4) {
                        dictionary.setInt32(name, bytes4ToUInt(valueBuf).toKotlinInt())
                    }
                    VdType.Int64 -> if (valueLen == 8) {
                        dictionary.setInt64(name, bytes64ToLong(valueBuf))
                    }
                    VdType.String -> dictionary.setString(name, String(valueBuf, UTF8Charset))
                    VdType.ByteArray -> dictionary.setByteArray(name, valueBuf)
                    else -> {
                    }
                }
            }
            return dictionary
        }

        @Throws(IOException::class)
        fun serialize(variantDictionary: VariantDictionary,
                      outputStream: OutputStream?) {
            if (outputStream == null) {
                return
            }
            outputStream.write2BytesUShort(VdVersion)
            for ((name, vd) in variantDictionary.dict) {
                val nameBuf = name.toByteArray(UTF8Charset)
                outputStream.writeByte(vd.type)
                outputStream.write4BytesUInt(UnsignedInt(nameBuf.size))
                outputStream.write(nameBuf)
                var buf: ByteArray
                when (vd.type) {
                    VdType.UInt32 -> {
                        outputStream.write4BytesUInt(UnsignedInt(4))
                        outputStream.write4BytesUInt(vd.value as UnsignedInt)
                    }
                    VdType.UInt64 -> {
                        outputStream.write4BytesUInt(UnsignedInt(8))
                        outputStream.write8BytesLong(vd.value as UnsignedLong)
                    }
                    VdType.Bool -> {
                        outputStream.write4BytesUInt(UnsignedInt(1))
                        outputStream.writeBooleanByte(vd.value as Boolean)
                    }
                    VdType.Int32 -> {
                        outputStream.write4BytesUInt(UnsignedInt(4))
                        outputStream.write4BytesUInt(UnsignedInt(vd.value as Int))
                    }
                    VdType.Int64 -> {
                        outputStream.write4BytesUInt(UnsignedInt(8))
                        outputStream.write8BytesLong(vd.value as Long)
                    }
                    VdType.String -> {
                        val value = vd.value as String
                        buf = value.toByteArray(UTF8Charset)
                        outputStream.write4BytesUInt(UnsignedInt(buf.size))
                        outputStream.write(buf)
                    }
                    VdType.ByteArray -> {
                        buf = vd.value as ByteArray
                        outputStream.write4BytesUInt(UnsignedInt(buf.size))
                        outputStream.write(buf)
                    }
                    else -> {
                    }
                }
            }
            outputStream.write(VdType.None.toInt())
        }
    }

    class VdType(val type: Byte, val value: Any) {

        companion object {
            const val None: Byte = 0x00
            const val UInt32: Byte = 0x04
            const val UInt64: Byte = 0x05
            const val Bool: Byte = 0x08
            const val Int32: Byte = 0x0C
            const val Int64: Byte = 0x0D
            const val String: Byte = 0x18
            const val ByteArray: Byte = 0x42
        }
    }
}